File: engine/src/main/java/de/hanno/hpengine/engine/graphics/buffer/AbstractPersistentMappedBuffer.java
package de.hanno.hpengine.engine.graphics.buffer;

import de.hanno.hpengine.engine.graphics.GpuContext;
import org.lwjgl.opengl.GL15;
import org.lwjgl.opengl.GL44;

import java.nio.ByteBuffer;
import java.nio.IntBuffer;

import static org.lwjgl.opengl.ARBBufferStorage.GL_MAP_COHERENT_BIT;
import static org.lwjgl.opengl.ARBBufferStorage.GL_MAP_PERSISTENT_BIT;
import static org.lwjgl.opengl.GL15.*;
import static org.lwjgl.opengl.GL30.GL_MAP_WRITE_BIT;

public abstract class AbstractPersistentMappedBuffer implements GPUBuffer {
    private final GpuContext gpuContext;
    protected int target = 0;
    private int id;
    protected volatile ByteBuffer buffer;
    private IntBuffer intBuffer;
    private boolean bound = false;

    public AbstractPersistentMappedBuffer(GpuContext gpuContext, int target) {
        this.gpuContext = gpuContext;
        this.target = target;
    }

    public ByteBuffer map(int requestedCapacityInBytes) {
        bind();
        if(requestedCapacityInBytes > buffer.capacity()){
            setCapacityInBytes(requestedCapacityInBytes);
        }

        buffer.clear();
        return buffer;
    }

    @Override
    public synchronized void setCapacityInBytes(int requestedCapacity) {
        int capacityInBytes = requestedCapacity;
        if(capacityInBytes <= 0) { capacityInBytes = 10; }

        if(buffer != null) {
            boolean needsResize = buffer.capacity()  <= capacityInBytes;
            if(needsResize) {
                gpuContext.execute(() -> {
                    bind();
                    if(GL15.glGetBufferParameteri(target, GL15.GL_BUFFER_MAPPED) == 1) {
                        glUnmapBuffer(target);
                    }
                    if(id > 0) {
                        glDeleteBuffers(id);
                        id = -1;
                    }
                    unbind();
                });
            } else {
                return;
            }
        }
        {
            int finalCapacityInBytes = 2*capacityInBytes;
            gpuContext.execute(() -> {
                bind();
                int flags = GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT;
                GL44.glBufferStorage(target, finalCapacityInBytes, flags);

                buffer = mapBuffer(finalCapacityInBytes, flags);
                intBuffer = buffer.asIntBuffer();
                unbind();
            });
        }
    }

    abstract protected ByteBuffer mapBuffer(long capacityInBytes, int flags);

    @Override
    public void bind() {
//        TODO: Make this somehow possible
//        if(bound) {return;}
        gpuContext.execute(bindBufferRunnable);
        bound = true;
    }

    private Runnable bindBufferRunnable = () -> {
        if (getId() <= 0) {
            setId(glGenBuffers());
        }
        glBindBuffer(target, getId());
    };

    @Override
    public void unbind() {
        gpuContext.execute(() -> glBindBuffer(target, 0));
        bound = false;
    }

    @Override
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    @Override
    public int getSizeInBytes() {
        return buffer.capacity();
    }

    @Override
    public void setSizeInBytes(int size) {
        setCapacityInBytes(size);
    }

    @Override
    public ByteBuffer getBuffer() {
        return buffer;
    }

    @Override
    public IntBuffer getIntBufferView() {
        return intBuffer;
    }

    public void dispose() {
        glDeleteBuffers(id);
    }
}

File: engine/src/main/java/de/hanno/hpengine/engine/graphics/buffer/AbstractPersistentMappedBuffer.kt
package de.hanno.hpengine.engine.graphics.buffer

import de.hanno.hpengine.engine.graphics.GpuContext
import org.lwjgl.BufferUtils
import org.lwjgl.opengl.GL44

import java.nio.ByteBuffer

import org.lwjgl.opengl.ARBBufferStorage.GL_MAP_COHERENT_BIT
import org.lwjgl.opengl.ARBBufferStorage.GL_MAP_PERSISTENT_BIT
import org.lwjgl.opengl.GL15.*
import org.lwjgl.opengl.GL30
import org.lwjgl.opengl.GL30.GL_MAP_WRITE_BIT
import org.lwjgl.opengl.GL45.glCopyNamedBufferSubData

val flags = GL_MAP_WRITE_BIT or GL_MAP_PERSISTENT_BIT or GL_MAP_COHERENT_BIT

abstract class AbstractPersistentMappedBuffer @JvmOverloads constructor(private val gpuContext: GpuContext<*>,
                                              protected var target: Int,
                                              capacityInBytes: Int = 1024) : GPUBuffer {
    private var bufferDefinition: BufferDefinition = createBuffer(capacityInBytes)
        set(value) {
            val oldBufferDefinition = bufferDefinition
            oldBufferDefinition.copyTo(value)
            field = value
            oldBufferDefinition.deleteBuffer()
        }

    override val id: Int
        get() = bufferDefinition.id
    override val buffer: ByteBuffer
        get() = bufferDefinition.buffer

    override var sizeInBytes: Int
        get() = buffer.capacity()
        set(value) {
            ensureCapacityInBytes(value)
        }

    private val bindBufferRunnable = {
        glBindBuffer(target, id)
    }

    @Synchronized
    override fun ensureCapacityInBytes(requestedCapacity: Int) {
        var capacityInBytes = requestedCapacity
        if (capacityInBytes <= 0) {
            capacityInBytes = 10
        }

        val needsResize = buffer.capacity() < capacityInBytes
        if (needsResize) {
            val newBufferDefinition = gpuContext.calculate {
                createBuffer(capacityInBytes)
            }
            bufferDefinition = newBufferDefinition
        }
    }

    private fun BufferDefinition.copyTo(newBuffer: BufferDefinition) {
        gpuContext.execute("copy buffers") {
            glCopyNamedBufferSubData(this.id, newBuffer.id, 0, 0, this.buffer.capacity().toLong())
        }
    }

    data class BufferDefinition(val id: Int, val buffer: ByteBuffer, val gpuContext: GpuContext<*>) {
        init {
            require(id > 0) { "Buffer id is invalid: $id" }
        }

        fun deleteBuffer() = gpuContext.execute("delete buffers") {
            glDeleteBuffers(id)
        }
    }
    private fun createBuffer(capacityInBytes: Int): BufferDefinition = gpuContext.calculate {
        val id = glGenBuffers()
        glBindBuffer(target, id)
        GL44.glBufferStorage(target, capacityInBytes.toLong(), flags)
        val xxxx = BufferUtils.createByteBuffer(capacityInBytes)
        val byteBuffer = GL30.glMapBufferRange(target,
                0, capacityInBytes.toLong(), flags,
//                null)!!
                xxxx)!!
        BufferDefinition(id, byteBuffer, gpuContext)
    }

    override fun bind() {
        gpuContext.execute("AbstractPersistentMappedBuffer.bind", bindBufferRunnable)
    }

    override fun unbind() {
        gpuContext.execute("AbstractPersistentMappedBuffer.unbind") { glBindBuffer(target, 0) }
    }
}
